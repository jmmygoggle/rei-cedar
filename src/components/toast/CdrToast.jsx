import clsx from 'clsx';
import propValidator from '../../utils/propValidator';
import BuildClass from '../../mixins/buildClass';
import IconXSm from '../icon/comps/x-sm';
import CdrButton from '../button/CdrButton';
import style from './styles/CdrToast.scss';

export default {
  name: 'CdrToast',
  components: {
    CdrButton,
    IconXSm,
  },
  mixins: [BuildClass],
  props: {
    type: {
      type: String,
      validator: (value) => propValidator(
        value,
        ['info', 'warning', 'success', 'error', 'default'],
      ),
      default: 'default',
    },
    open: {
      type: Boolean,
      default: false,
    },
    autoDismiss: {
      type: Boolean,
      default: true,
    },
    dismissDelay: {
      type: Number,
      default: 5000,
    },
  },
  data() {
    return {
      style,
      closed: null,
      opened: null,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-toast';
    },
    borderClass() {
      return this.modifyClassName(this.baseClass, this.type);
    },
    iconClass() {
      return this.modifyClassName(`${this.baseClass}__icon-left`, this.type);
    },
    themeClass() {
      return this.modifyClassName(`${this.baseClass}__main`, this.type);
    },
  },
  watch: {
    open(newValue, oldValue) {
      if (!!newValue === !!oldValue) return;
      if (newValue) {
        this.openToast();
      } else {
        this.closeToast();
      }
    },
  },
  methods: {
    openToast() {
      this.closed = false;
      this.opened = true;
      if (this.autoDismiss) {
        setTimeout(this.closeToast, this.dismissDelay);
      }
    },
    closeToast() {
      this.opened = false;
      this.closed = true;
    },

  },
  render() {
    return (
      <transition
        enter-class={clsx(this.style['cdr-toast--toast-enter'])}
        enter-active-class={clsx(this.style['cdr-toast--toast-enter-active'])}
        leave-to-class={clsx(this.style['cdr-toast--toast-leave-to'])}
        leave-active-class={clsx(this.style['cdr-toast--toast-leave-active'])}
      >
        { this.opened
          ? <div
              class={clsx(
                this.style[this.baseClass],
                this.borderClass,
              )}
            >
          <div class={clsx(this.style['cdr-toast__main'], this.themeClass)}>
            { this.$slots['icon-left'] && (
                  <div class={clsx(this.style['cdr-toast__icon-left'], this.iconClass)}>
                    {this.$slots['icon-left'] }
                  </div>
            )}
            <span class={this.style['cdr-toast__message']}>
              { this.$slots.default }
            </span>
            <cdr-button
                class={this.style['cdr-toast__close-button']}
                icon-only
                on-click={ this.closeToast }
                aria-label="Close"
                size="small"
              >
                <icon-x-sm
                  slot="icon"
                  inherit-color
                />
              </cdr-button>
          </div>
        </div>
          : null }
    </transition>
    );
  },
};
