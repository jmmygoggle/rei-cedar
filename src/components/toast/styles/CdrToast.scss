@import '../../../css/settings/index.scss';
@import './vars/CdrToast.vars.scss';

.cdr-toast {
  @include cdr-toast-base-mixin;

  &__transition {
    &--toast-enter {
      opacity: 0;
      transform: translateX(2.4rem);
    }
    &--toast-enter-active {
      transition: all $cdr-duration-5-x $timing-function;
    }
    &--toast-leave-to {
      opacity: 0;
    }
    &--toast-leave-active {
      transition: opacity $cdr-duration-2-x $timing-function;
    }
  }

  &__main {
    display: grid;
    grid-template-columns: auto 1fr auto;
    border-top-right-radius: $cdr-radius-softer;
    border-bottom-right-radius: $cdr-radius-softer;
    grid-template-areas: "icon-left message close-button";
  }

  &__icon-left {
    grid-area: icon-left;
    display: inherit;
    align-items: center;
    justify-items: center;
    & svg {
      width: 3.2rem;
      stroke: white;
      stroke-width: 0.4rem;
      paint-order: stroke fill;
      margin: $cdr-space-quarter-x;
  
    }
  }

  &__message {
    grid-area: message;
    padding: $cdr-space-half-x;
  }

  &__close-button {
    grid-area: close-button;
    height: min-content;
    & svg {
      fill: $cdr-color-text-emphasis!important;
    }
  }

  &--default {
    & .cdr-toast__main {
      @include cdr-toast-default-mixin;
      & .cdr-toast__icon-left {
          background-color: $cdr-color-background-message-default-02;
        }
      }
    border-left-color: $cdr-color-border-message-default-01;
  }

  &--info {
    & .cdr-toast__main {
      @include cdr-toast-info-mixin;
      & .cdr-toast__icon-left {
        background-color: $cdr-color-background-message-info-02;
      }
    }
    border-left-color: $cdr-color-border-message-info-01;
  }

  &--warning {
    & .cdr-toast__main {
      @include cdr-toast-warning-mixin;
      & .cdr-toast__icon-left {
        background-color: $cdr-color-background-message-warning-02;
      }
    }
    border-left-color: $cdr-color-border-message-warning-01;
  }

  &--success {
    & .cdr-toast__main {
      @include cdr-toast-success-mixin;
      & .cdr-toast__icon-left {
        background-color: $cdr-color-background-message-success-02;
      }
    }
    border-left-color: $cdr-color-border-message-success-01;
  }

  &--error {
    & .cdr-toast__main {
      @include cdr-toast-error-mixin;
      & .cdr-toast__icon-left {
        background-color: $cdr-color-background-message-error-02;
      }
    }
    border-left-color: $cdr-color-border-message-error-01;
  }
}