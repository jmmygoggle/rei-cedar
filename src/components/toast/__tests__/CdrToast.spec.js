import { mount } from '@vue/test-utils';
import CdrToast from 'componentdir/toast/CdrToast';
import IconCheckFill from 'componentdir/icon/comps/check-fill';

describe('CdrToast', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(CdrToast, {
      slots: {
        default: 'toast content',
        'icon-left': IconCheckFill,
      },
    });
  });
  it('matches snapshot', async () => {
    wrapper.setProps({ open: true });
    wrapper.setData({ opened: true });
    await wrapper.vm.$nextTick(() => {
      expect(wrapper.element).toMatchSnapshot();
    })
  });
  it('watches for close triggers', async () => {
    wrapper.setProps({ open: false });
    await wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.closeToast).toBeCalled();
    });
  });
});