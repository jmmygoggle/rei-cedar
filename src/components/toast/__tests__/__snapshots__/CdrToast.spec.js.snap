// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrToast matches snapshot 1`] = `
<transition-stub
  enteractiveclass="cdr-toast--toast-enter-active"
  enterclass="cdr-toast--toast-enter"
  leaveactiveclass="cdr-toast--toast-leave-active"
  leavetoclass="cdr-toast--toast-leave-to"
>
  <div
    class="cdr-toast cdr-toast--default"
  >
    <div
      class="cdr-toast__main cdr-toast__main--default"
    >
      <div
        class="cdr-toast__icon-left cdr-toast__icon-left--default"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm4.77-13.357a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 00-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
            role="presentation"
          />
        </svg>
      </div>
      <span
        class="cdr-toast__message"
      >
        toast content
      </span>
      <button
        aria-label="Close"
        class="cdr-button cdr-button--primary cdr-button--icon-only cdr-toast__close-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="cdr-icon cdr-icon--inherit-color"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13.406 12.006l3.297-3.296a1 1 0 10-1.414-1.414l-3.297 3.295-3.285-3.295A1 1 0 107.293 8.71l3.285 3.295-3.285 3.288a1 1 0 001.414 1.415l3.285-3.289 3.297 3.289a1 1 0 001.414-1.415l-3.297-3.287z"
            role="presentation"
          />
        </svg>
      </button>
    </div>
  </div>
</transition-stub>
`;
